### PR TITLE
API: Changes loss graph API to work off of AbstractLossGraph objects

### PR DESCRIPTION
**Pull request by jfkirk**
This one's a doozy, and lays the groundwork for loss functions with random sampling.

1. Changes loss_graph top level API to take in AbstractLossGraph objects.
2. Modifies all implemented loss graphs to inherit AbstractLossGraph.
3. Modifies graph construction to only pass serial nodes to loss function when loss function isn't dense for safety.
4. Updates README to show AbstractLossGraph example.
5. Moves readme test cases to their own module to test imports inside the unit tests.
6. Adds item sampler stub (this was intended for a later PR, but whoops).
7. Adds RMSEDenseLossGraph -- a version of RMSE that treats items with no interaction as interactions of value `0`.

### CODE REVIEW

**Comment by jfkirk** [comment]
cc @mesanders

File: tensorrec/loss_graphs.py [hunk 1]
```diff
@@ -1,93 +1,102 @@
...
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, is_dense=False):
+        self.is_dense = is_dense
+
+    @abc.abstractmethod
+    def loss_graph(self, tf_prediction_serial, tf_interactions_serial, tf_prediction, tf_interactions, tf_rankings,
+                   tf_alignment):
+        """
+        TODO document
```
Comment (jfkirk) [inline tensorrec/loss_graphs.py]: Oops -- slipped by me.
Won't merge until this is documented.

File: tensorrec/tensorrec.py [hunk 2]
```diff
@@ -282,13 +284,23 @@ def _build_tf_graph(self, n_user_features, n_item_features):
...
-                                                     tf_prediction=self.tf_prediction,
-                                                     tf_interactions=tf_interactions,
-                                                     tf_rankings=self.tf_rankings,
-                                                     tf_alignment=self.tf_alignment)
+        # Compose loss function args
+        # This composition is for execution safety: it prevents loss functions that are incorrectly configured from
+        # having visibility of certain nodes.
+        loss_graph_kwargs = {
+            'tf_prediction_serial': self.tf_prediction_serial,
+            'tf_interactions_serial': tf_interactions_serial,
```
Comment (mesanders) [inline tensorrec/tensorrec.py]: Is it on purpose that one of the values above calls self and the other doesn't?

File: tensorrec/tensorrec.py [hunk 3]
```diff
@@ -17,7 +17,7 @@ class TensorRec(object):
     def __init__(self, n_components=100,
                  user_repr_graph=linear_representation_graph,
                  item_repr_graph=linear_representation_graph,
-                 loss_graph=rmse_loss,
+                 loss_graph=RMSELossGraph(),
```
Comment (mesanders) [inline tensorrec/tensorrec.py]: Update documentation below for 

> :param loss_graph: Method

Comment (jfkirk) [reply tensorrec/tensorrec.py]: Good catch!

File: tensorrec/tensorrec.py (hunk 2 above)
Comment (jfkirk) [reply tensorrec/tensorrec.py]: Yes -- the ones with `self.` are graph hooks that the TensorRec object keeps track of. The rest are not used outside of the graph construction method.

**Pull request merged**